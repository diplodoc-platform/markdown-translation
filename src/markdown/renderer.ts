import MarkdownIt from 'markdown-it';
import {
    mdRenderer,
    MarkdownRendererParams,
    MarkdownRendererEnv,
} from '@diplodoc/markdown-it-markdown-renderer';

// configure with diplodoc plugins
// @ts-ignore
import meta from 'markdown-it-meta';
import notes from '@doc-tools/transform/lib/plugins/notes';

import markdownHandlers, {MarkdownHandlersState} from './handlers';
import hooks, {HooksParameters} from './hooks';

export type MarkdownRendererState = MarkdownHandlersState;

export type RenderParameters = {
    skeleton: string;
    translations: Map<string, string>;
} & DiplodocParameters;

export type DiplodocParameters = {
    lang?: string;
};

function render(parameters: RenderParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const md = new MarkdownIt('commonmark', {html: true}) as HooksParameters['markdownit'];

    const {handlers, initState} = markdownHandlers.generate(parameters);
    const markdownHooks = hooks.generate({markdownit: md});
    const mdOptions: MarkdownRendererParams<MarkdownHandlersState> = {
        handlers,
        initState,
        hooks: markdownHooks.hooks,
    };
    const diplodocOptions = {
        lang: parameters.lang ?? 'ru',
    };

    const env: MarkdownRendererEnv = {
        source: parameters.skeleton.split('\n'),
    };

    md.use(mdRenderer, mdOptions);

    // diplodoc plugins
    md.use(meta);
    md.use(notes, diplodocOptions);

    return md.render(parameters.skeleton, env);
}

function validParameters(parameters: RenderParameters) {
    const {skeleton, translations} = parameters;

    const skeletonCondition = skeleton !== undefined;
    const translationsCondition = translations !== undefined;

    const conditions = [
        skeletonCondition,
        translationsCondition,
        validDiplodocParameters(parameters),
    ];

    return conditions.reduce((a, v) => a && v, true);
}

function validDiplodocParameters(parameters: DiplodocParameters) {
    const {lang} = parameters;

    return lang === undefined || lang === 'ru' || lang === 'en';
}

export {render, validParameters, validDiplodocParameters};
export default {render, validParameters, validDiplodocParameters};
