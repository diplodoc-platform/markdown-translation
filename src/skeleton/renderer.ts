import MarkdownIt from 'markdown-it';
import {
    mdRenderer,
    MarkdownRendererParams,
    MarkdownRendererEnv,
} from '@diplodoc/markdown-it-markdown-renderer';

// configure with diplodoc plugins
// @ts-ignore
import meta from 'markdown-it-meta';
// @ts-ignore
import sup from 'markdown-it-sup';
import notes from '@doc-tools/transform/lib/plugins/notes';
import cut from '@doc-tools/transform/lib/plugins/cut';
import checkbox from '@doc-tools/transform/lib/plugins/checkbox';

import skeletonHandlers, {SkeletonHandlersState} from './handlers';
import hooks, {HooksParameters} from './hooks';

export type SkeletonRendererState = SkeletonHandlersState;

export type RenderParameters = {
    markdown: string;
} & DiplodocParameters;

export type DiplodocParameters = {
    lang?: string;
};

function render(parameters: RenderParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const {markdown, lang} = parameters;

    const md = new MarkdownIt({html: true}) as HooksParameters['markdownit'];
    const env: MarkdownRendererEnv = {source: markdown.split('\n')};

    const {handlers, initState} = skeletonHandlers.generate();

    const skeletonHooks = hooks.generate({markdownit: md});
    const mdOptions: MarkdownRendererParams<SkeletonHandlersState> = {
        handlers,
        initState,
        hooks: skeletonHooks.hooks,
    };
    const diplodocOptions = {
        lang: lang ?? 'ru',
    };

    md.use(mdRenderer, mdOptions);

    // diplodoc plugins
    md.use(meta, diplodocOptions);
    md.use(notes, diplodocOptions);
    md.use(cut, diplodocOptions);
    md.use(sup, diplodocOptions);
    md.use(checkbox, diplodocOptions);

    return md.render(markdown, env);
}

function validParameters(parameters: RenderParameters) {
    const {markdown, lang} = parameters;

    const markdownCondition = markdown !== undefined;
    const langCondition = lang === undefined || lang === 'ru' || lang === 'en';
    const conditions = [markdownCondition, langCondition];

    return conditions.reduce((a, v) => a && v, true);
}

export {render, validParameters};
export default {render, validParameters};
