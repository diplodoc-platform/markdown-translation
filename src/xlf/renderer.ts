import MarkdownIt from 'markdown-it';
import {customRenderer, CustomRendererParams} from '@diplodoc/markdown-it-custom-renderer';

// configure with diplodoc plugins
// @ts-ignore
import meta from 'markdown-it-meta';
import notes from '@doc-tools/transform/lib/plugins/notes';

import {template} from './generator';
import rules, {XLFRulesState} from './rules';
import hooks, {HooksParameters} from './hooks';

export type XLFRendererState = XLFRulesState;

export type RenderParameters = {
    markdown: string;
} & template.TemplateParameters &
    DiplodocParameters;

export type DiplodocParameters = {
    lang?: string;
};

function render(parameters: RenderParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const wrapper = template.generate(parameters);

    const xlfRenderer = new MarkdownIt('commonmark', {html: true}) as HooksParameters['markdownit'];
    const xlfRules = rules.generate(wrapper);
    const xlfHooks = hooks.generate({template: wrapper.template, markdownit: xlfRenderer});
    const xlfOptions: CustomRendererParams<XLFRendererState> = {
        rules: xlfRules.rules,
        hooks: xlfHooks.hooks,
        initState: xlfRules.initState,
    };
    const diplodocOptions = {
        lang: parameters.lang ?? 'ru',
    };

    xlfRenderer.use(customRenderer, xlfOptions);

    // diplodoc plugins
    xlfRenderer.use(meta);
    xlfRenderer.use(notes, diplodocOptions);

    return xlfRenderer.render(parameters.markdown);
}

function validParameters(parameters: RenderParameters) {
    const {lang, markdown} = parameters;

    const markdownCondition = markdown !== undefined;
    const langCondition = lang === undefined || lang === 'ru' || lang === 'en';
    const conditions = [template.validParameters(parameters), markdownCondition, langCondition];

    return conditions.reduce((a, v) => a && v, true);
}

export {render, validParameters};
export default {render, validParameters};
