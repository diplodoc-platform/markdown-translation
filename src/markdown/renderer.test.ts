import {render, RenderParameters} from './renderer';
import {
    skeleton,
    skeletonWithVariables,
    translations,
    translationsWithVariables,
    skeletonWithConditionals,
    translationsWithConditionals,
    skeletonWithLoops,
    translationsWithLoops,
    skeletonWithFilters,
    translationsWithFilters,
    skeletonWithFunctions,
    translationsWithFunctions,
    skeletonWithMeta,
    translationsWithMeta,
} from 'src/__fixtures__';

describe('smoke', () => {
    test('it works', () => {
        const parameters = {
            skeleton: '',
            translations: new Map<string, string>(),
        };

        render(parameters);
    });
});

describe('validates parameters', () => {
    it('works with valid parameters', () => {
        const parameters = {
            skeleton: '',
            translations: new Map<string, string>(),
        };

        render(parameters);
    });

    it('throws on invalid parameters', () => {
        const parameters = {} as RenderParameters;
        expect(() => render(parameters)).toThrow();
    });
});

describe('markdown rendering', () => {
    it('renders identital markdown without translations', () => {
        const parameters = {
            skeleton: '# title\nhello, world\n',
            translations: new Map<string, string>([
                ['1', 'титул'],
                ['2', 'привет, мир'],
            ]),
        };

        const generated = render(parameters);

        expect(generated).toMatchSnapshot();
    });

    it('renders translated text instead of hashes', () => {
        const parameters = {
            skeleton,
            translations,
        };

        const generated = render(parameters);
        expect(generated).toMatchSnapshot();
    });

    it('renders translated text instead of hashes, with liquid variables in text', () => {
        const parameters = {
            skeleton: skeletonWithVariables,
            translations: translationsWithVariables,
        };

        const generated = render(parameters);
        expect(generated).toMatchSnapshot();
    });

    it('renders translated text instead of hashes, with liquid filters in text', () => {
        const parameters = {
            skeleton: skeletonWithFilters,
            translations: translationsWithFilters,
        };

        const generated = render(parameters);
        expect(generated).toMatchSnapshot();
    });

    it('renders translated text instead of hashes, with liquid functions in text', () => {
        const parameters = {
            skeleton: skeletonWithFunctions,
            translations: translationsWithFunctions,
        };

        const generated = render(parameters);
        expect(generated).toMatchSnapshot();
    });

    it('renders translated text instead of hashes, with liquid conditionals in text', () => {
        const parameters = {
            skeleton: skeletonWithConditionals,
            translations: translationsWithConditionals,
        };

        const generated = render(parameters);
        expect(generated).toMatchSnapshot();
    });

    it('renders translated txt instead of hashes, with liquid loops in text', () => {
        const parameters = {
            skeleton: skeletonWithLoops,
            translations: translationsWithLoops,
        };

        const generated = render(parameters);
        expect(generated).toMatchSnapshot();
    });

    it('renders translated text instead of hashes, with meta in text', () => {
        const parameters = {
            skeleton: skeletonWithMeta,
            translations: translationsWithMeta,
        };

        const generated = render(parameters);
        expect(generated).toMatchSnapshot();
    });
});
