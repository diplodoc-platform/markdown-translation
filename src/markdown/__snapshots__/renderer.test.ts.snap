// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown rendering renders identital markdown without translations 1`] = `
"# title
hello, world"
`;

exports[`markdown rendering renders translated text instead of hashes 1`] = `
"# Heading 1
Paragraph
- First list item
- Second list item
Heading 2
-----------
Paragraph with **bold text** and *italics*

Also key features of the project structure highlight extremely interesting features of the picture as a whole. Sentences add up to paragraphs — and you will enjoy another delirium.
> text inside of the blockquote
1. First ordered list item
2. Second ordered list item

[Link text. Link text continued.](files/file.md \\"Link description. Link description continued.\\")

![Alternative image text Alt. text continued](files/image.png \\"Image title Image title continued\\")"
`;

exports[`markdown rendering renders translated text instead of hashes, with anchors syntax in text 1`] = `
"# heading with implicit anchor
paragraph
## heading with explicit anchor {#anchor}"
`;

exports[`markdown rendering renders translated text instead of hashes, with checkbox syntax in text 1`] = `
"[ ] first

[x] second
- [_] to be done underscore
- [-] to be done dash
- [ ] to be done space
- [x] done cross
- [X] done capital cross

- list element
  - [x] done
  - [ ] to be done"
`;

exports[`markdown rendering renders translated text instead of hashes, with cuts in text 1`] = `
"
{% cut \\"Cut Title\\" %}

Content inside of the cut

{% endcut %}"
`;

exports[`markdown rendering renders translated text instead of hashes, with gfm tables in text 1`] = `
"|Header 1|Header 2|
|-----------|-----------|
|[link text](https://www.google.com)|**bold text**|
|![image text](image.png \\"hint\\")|*italics*|
"
`;

exports[`markdown rendering renders translated text instead of hashes, with liquid conditionals in text 1`] = `
"# Титул Документа
{% if OS == 'iOS' %}

Скачайте приложение из  [{{ ios-marketplace }} Стора](https://www.apple.com/ios/app-store).

{% else %}

Скачайте приложение из  [{{ other-marketplace }} Стора](https://play.google.com).

Какой-то текст {% if  OS == 'iOS' %} Apple {% else %} Android {% endif %} Продолжение текста.

{% endif %}"
`;

exports[`markdown rendering renders translated text instead of hashes, with liquid filters in text 1`] = `
"Титул Документа
Привет {{ user.name | capitalize }}!

всего пользователей: {{ users | length }}"
`;

exports[`markdown rendering renders translated text instead of hashes, with liquid functions in text 1`] = `
"Титул Документа
Привет П{{ user.name.slice(1) }}!

Привет П{{ user.name.slice(1, 2) }}вел!"
`;

exports[`markdown rendering renders translated text instead of hashes, with liquid variables in text 1`] = `
"# {{ service-name }} Сервис
Попробуйте {{ service-name }} и расскажите нам что вы думаете о нём.

Также присмотритесь к [{{ other-service-name }}](files/file.md \\"{{ other-service-name }} описание\\")

![скриншот использования {{ service-name }}](files/image.png \\"{{ service-name }} в действие\\")"
`;

exports[`markdown rendering renders translated text instead of hashes, with meta in text 1`] = `
"---
title: Creation of the Windows virtual machine
description: >-
  Create Windows virtual machine with {{ compute-short-name }} service. You can
  connect to the virtual machine using Remote Desktop Protocol (RDP). Before
  connecting ensure that NLA is enabled.
key: |
  Sammy Sosa

  Let's Goooo
seq:
  - windows
  - windows vm
  - windows vm
  - windows virtual machine
  - virtual machine
  - vm
seq_of_mappings:
  - name: Mark
    age: 13
  - name: Noam
    age: 99
object_inline:
  name: John Smith
  age: 33
array_inline:
  - milk
  - pumpkin pie
  - eggs
  - juice
seq_of_seq:
  - - Mark McGwire
    - 65
    - 0.278
  - - Sammy Sosa
    - 63
    - 0.288
nesting:
  - children:
      - name: bob
      - name: marry
---
# Meta
Comments and metadata - not displayed in the final file. They are used to make comments for other writers or set meta into final html for seo."
`;

exports[`markdown rendering renders translated text instead of hashes, with notes in text 1`] = `
"
{% note info \\"Information\\" %}

information

{% endnote %}
- list with note inside

  {% note alert \\"Alert\\" %}

  alert

  {% endnote %}"
`;

exports[`markdown rendering renders translated text instead of hashes, with sup syntax in text 1`] = `
"paragraph of text 1^st^ line
this is 2^nd^ line
[1^st^ link inside the paragraph](file.md)"
`;

exports[`markdown rendering renders translated txt instead of hashes, with liquid loops in text 1`] = `
"# Титул Документа
{% for element in array %}

Какой-то текст {{ variable_name }} продолжение текста.

{% endfor %}

Перед однострочным циклом {% for element in array %} текст внутри содержащий {{variable}} {% endfor %} после однострочного цикла"
`;
